File: hto/60/hto_60.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>
#include <math.h>
#include <stdint.h>
#include <string.h>

double* twoSum(int n, int* returnSize){
    int width = 6 * n + 1;
    int dp[12][100] = {0};
    for (int i = 0; i <= 6; i++) {
        dp[1][i] = 1;
    }
    for (int i = 2; i <= n; i++) {
        for (int j = i; j <= 6*i; j++) {
            for (int cur = 1; cur <= 6; cur++) {
                if (j - cur <= 0) break;
                dp[i][j] += dp[i-1][j-cur];
            }
        }
    }
    // double all = pow(6.0, (double)n);
    int m = n;
    double all = 1.;
    while (m--)  all *= 6.;
    double *res = (double *)malloc(sizeof(double) * width);
    int j = 0;
    for (int i = n; i <= 6 * n; i++) {
        res[j++] = dp[n][i] / all;
    }
    *returnSize = j;
    return res;
}


int main(void) {
    int x = 0;
    double *res = twoSum(2, &x);
    while (*res != 0) {
        printf("%lf ", *res);
        res++;
    }
}

File: hto/60/hto_60.java
class Solution {
    public double[] twoSum(int n) {
        double[] res = new double[5*n+1];
        int[] dp = new int[6*n+1];
        for (int i = 1; i <= 6; i++) {
            dp[i] = 1;
        }

        for (int i = 2; i <= n; i++) {
            for (int j = 6*i; j >= i; j--) {
                dp[j] = 0;
                for (int k = 1; k <= 6; k++) {
                    if (j - k >= i - 1)
                        dp[j] += dp[j-k];
                }
            }
        }
        double denominator = Math.pow(6, n);
        for (int i = n; i <= 6*n; i++) {
            res[i-n] = dp[i] / denominator;
        }
        return res;
    }
}

